Scroll To Text Fragments fragments should alwasy be scrolled to the cente of the screen.

megangardner opened the pull request:
#### 6ae0bcb1c73ce6ea808e09557db62a0dbc5446d6
<pre>
Scroll To Text Fragments fragments should alwasy be scrolled to the cente of the screen.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245169">https://bugs.webkit.org/show_bug.cgi?id=245169</a>
&lt;rdar://99904303&gt;

Reviewed by Tim Horton.

Our code that scrolls selections into visible does not scroll them if they are already visible
at the current page scroll. We are wanting this to always go to the center so that it is easier
to hold the position, rather than have the scrolled text jump around a bit as additional content
is loaded in.

* Source/WebCore/editing/Editor.cpp:
(WebCore::TemporarySelectionChange::setSelection):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelection):
(WebCore::FrameSelection::updateAndRevealSelection):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::textFragmentIndicatorTimerFired):

Canonical link: <a href="https://commits.webkit.org/254477@main">https://commits.webkit.org/254477@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d90533e82ba16f204c7f987f6716a812b6ac518e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98471 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154783 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93162 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32222 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27778 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81517 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92943 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25589 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76086 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25526 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80427 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68500 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30004 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14499 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29730 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15461 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3139 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33176 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38418 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34554 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/editing/Editor.h:
```diff
@@ -128,6 +128,8 @@ enum class TemporarySelectionOption : uint8_t {
 
     UserTriggered = 1 << 7,
+    
+    ForceCenterScroll = 1 << 8,
```
Fine, but would be nice if this used the wording from ScrollAlignment. *Ideally* we'd just pass a ScrollAlignment in, but for some reason we've (previously) gotten really hung up on just having flags here.

megangardner committed:
Scroll To Text Fragments fragments should alwasy be scrolled to the cente of the screen.
https://bugs.webkit.org/show_bug.cgi?id=245169
<rdar://99904303>

Reviewed by Tim Horton.

Our code that scrolls selections into visible does not scroll them if they are already visible
at the current page scroll. We are wanting this to always go to the center so that it is easier
to hold the position, rather than have the scrolled text jump around a bit as additional content
is loaded in.

* Source/WebCore/editing/Editor.cpp:
(WebCore::TemporarySelectionChange::setSelection):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelection):
(WebCore::FrameSelection::updateAndRevealSelection):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::textFragmentIndicatorTimerFired):

Canonical link: https://commits.webkit.org/254477@main
